UnitySDK/Assets/ML-Agents/Scripts/InferenceBrain/TensorGenerator.cs

```diff
@@ -78,13 +76,41 @@ public interface IGenerator
             m_Dict[TensorNames.ValueEstimateOutput] = new BiDimensionalOutputGenerator(allocator);
         }
 
-        public void InitializeVisualObservations(Agent agent, ITensorAllocator allocator)
+        public void InitializeObservations(Agent agent, ITensorAllocator allocator)
         {
-            for (var visIndex = 0; visIndex < agent.sensors.Count; visIndex++)
+            // Loop through the sensors on a representative agent.
+            // For vector observations, add the index to the (single) VectorObservationGenerator
+            // For visual observations, make a VisualObservationInputGenerator
+            var visIndex = 0;
+            var vecIndex = 0;
+            VectorObservationGenerator vecObsGen = null;
+            for (var sensorIndex = 0; sensorIndex < agent.sensors.Count; sensorIndex++)
             {
-                // TODO handle non-visual Sensors too - need to index better
-                m_Dict[TensorNames.VisualObservationPlaceholderPrefix + visIndex] =
-                    new VisualObservationInputGenerator(visIndex, allocator);
+                var sensor = agent.sensors[sensorIndex];
+                var shape = sensor.GetFloatObservationShape();
+                // TODO generalize - we currently only have vector or visual, but can't handle "2D" observations
+                var isVectorSensor = (shape.Length == 1);
+                if (isVectorSensor)
+                {
+                    if (vecObsGen == null)
+                    {
+                        vecObsGen = new VectorObservationGenerator(allocator);
+                    }
+
+                    vecObsGen.AddSensorIndex(sensorIndex);
+                    vecIndex++;
```

Comment:
is this used anywhere?

Reply:
is it meant to be used above? in `vecObsGen.AddSensorIndex`?

Reply:
Not used, I should just remove it.